src/charm.py

```diff
@@ -0,0 +1,270 @@
+                ServicePort(name="sctp-grpc", port=SCTP_GRPC_PORT),
+            ],
+        )
+        self._amf_database = DatabaseRequires(
+            self, relation_name="amf-database", database_name=AMF_DATABASE_NAME
+        )
+        self._default_database = DatabaseRequires(
+            self, relation_name="default-database", database_name=DEFAULT_DATABASE_NAME
+        )
+
+        self.framework.observe(
+            self.on.default_database_relation_joined,
+            self._on_amf_pebble_ready,
+        )
+        self.framework.observe(
+            self.on.amf_database_relation_joined,
+            self._on_amf_pebble_ready,
+        )
+        self.framework.observe(
+            self._default_database.on.database_created, self._on_amf_pebble_ready
+        )
+        self.framework.observe(self._amf_database.on.database_created, self._on_amf_pebble_ready)
+        self.framework.observe(self.on.amf_pebble_ready, self._on_amf_pebble_ready)
+        self.framework.observe(self._nrf_requires.on.nrf_available, self._on_amf_pebble_ready)
+
+    def _on_amf_pebble_ready(
+        self,
+        event: EventBase,
+    ) -> None:
+        """Handle pebble ready event for AMF container.
+
+        Args:
+            event (PebbleReadyEvent, DatabaseCreatedEvent, NRFAvailableEvent): Juju event
+        """
+        if not self._amf_container.can_connect():
+            self.unit.status = MaintenanceStatus("Waiting for service to start")
+            event.defer()
+            return
+        for relation in ["fiveg_nrf", "amf-database", "default-database"]:
+            if not self._relation_created(relation):
+                self.unit.status = BlockedStatus(f"Waiting for {relation} relation")
+                return
+        if not self._default_database_is_available:
+            self.unit.status = WaitingStatus("Waiting for the default database to be available")
+            event.defer()
+            return
+        if not self._amf_database_is_available:
+            self.unit.status = WaitingStatus("Waiting for the amf database to be available")
+            event.defer()
+            return
+        if not self._default_database_info:
+            self.unit.status = WaitingStatus("Waiting for default database info to be available")
+            event.defer()
+            return
+        if not self._nrf_data_is_available:
+            self.unit.status = WaitingStatus("Waiting for NRF data to be available")
+            event.defer()
+            return
+        if not self._config_file_is_pushed:
+            self._push_config_file(
+                database_url=self._default_database_info["uris"].split(",")[0],
+                nrf_url=self._nrf_requires.nrf_url,
+            )
+        self._amf_container.add_layer("amf", self._amf_pebble_layer, combine=True)
+        self.unit.status = ActiveStatus()
+
+    def _push_config_file(self, database_url: str, nrf_url: str) -> None:
```

Comment:
It'd be cleaner for this method to accept all the parameters that it will place in the `render` method. Right now `render` uses a mix of environment variables, class properties and parameters.

Reply:
✅